test/org/github/jamm/strategies/MemoryMeterStrategyTest.java

```diff
@@ -0,0 +1,693 @@
+        assertEquals(expected, m2.measureDeep(empty));
+        assertEquals(expected, m2.measureDeep(readOnlyEmpty));
+
+        expected = sizeShallowEmptyBuffer + emptyOne.remaining();
+        assertEquals("Deep duplicated 1-byte ByteBuffer", expected, m2.measureDeep(emptyOne));
+        assertEquals("Deep duplicated 1-byte ByteBuffer", expected, m2.measureDeep(readOnlyEmptyOne));
+
+        expected = sizeShallowEmptyBuffer + reference.measureArray(twenty.array());
+        assertEquals("Twenty bytes ByteBuffer", expected, m2.measureDeep(twenty)); // The buffer represent the full array therefore nothing is shared
+        assertEquals("Twenty bytes ByteBuffer", expected, m2.measureDeep(readOnlyTwenty));
+
+        expected = sizeShallowEmptyBuffer + five.remaining();
+        assertEquals("Five bytes ByteBuffer", expected, m2.measureDeep(five));
+        assertEquals("Five bytes ByteBuffer", expected, m2.measureDeep(readOnlyFive));
+    }
+
+    @Test
+    public void testDirectByteBuffer() {
+
+        ByteBuffer empty = ByteBuffer.allocateDirect(0);
+        ByteBuffer readOnlyEmpty = empty.asReadOnlyBuffer();
+        ByteBuffer one = ByteBuffer.allocateDirect(1);
+        ByteBuffer readOnlyOne = one.asReadOnlyBuffer();
+        ByteBuffer emptyOne = (ByteBuffer) one.duplicate().position(1);
+        ByteBuffer readOnlyEmptyOne = emptyOne.asReadOnlyBuffer();
+        ByteBuffer twenty = ByteBuffer.allocateDirect(20);
+        ByteBuffer readOnlyTwenty = twenty.asReadOnlyBuffer();
+        ByteBuffer five = (ByteBuffer) twenty.slice().limit(5);
+        ByteBuffer readOnlyFive = five.asReadOnlyBuffer();
+
+        MemoryMeter m1 = MemoryMeter.builder().withGuessing(guess).build();
+
+        // Read-only and normal direct ByteBuffer have the same fields and therefore the same size
+        long sizeShallowBuffer = reference.measure(empty);
+        assertEquals("empty ByteBuffer", sizeShallowBuffer, m1.measure(empty));
+        assertEquals("empty ByteBuffer", sizeShallowBuffer, m1.measure(readOnlyEmpty));
+
+        // Measure deep will include reference to the cleaner for Read-write buffers
+        //
+        // root [java.nio.DirectByteBuffer] 136 bytes (64 bytes)
+        //   |
+        //   +--cleaner [jdk.internal.ref.Cleaner] 72 bytes (40 bytes)
+        //     |
+        //     +--thunk [java.nio.DirectByteBuffer$Deallocator] 32 bytes (32 bytes)
+        long sizeDeepRwBuffer = reference.measureDeep(empty);
+        assertEquals("Deep empty ByteBuffer", sizeDeepRwBuffer, m1.measureDeep(empty));
+        assertEquals("Deep 1-byte ByteBuffer", sizeDeepRwBuffer, m1.measureDeep(one));
+        assertEquals("Twenty bytes ByteBuffer", sizeDeepRwBuffer, m1.measureDeep(twenty));
+
+        // If a DirectByteBuffer is referencing a part of another DirectByteBuffer (read only or slice buffer), it will have a 
+        // reference to the original buffer through the att (attachement) field and no cleaner
```

Comment:
the attachment but only pre-12

Reply:
What changed with Java 12 is that before you could have a chain of references which is avoided since Java 12. I mentioned it bellow in another comment.